Kazaz/update oz versions

kazazor opened the pull request:
## What's the purpose of your PR?

- [ ] Fixing a bug
- [ ] A new Feature
- [X] Techdebt
- [ ] General Task

## What your PR is all about?
* In the BDStable contract, I had to remove all the SafeMath usage as it was never explicitly used in the contract.
* Updated the OZ version
* Removed the use of the 0.5.0 version of UniswapPairOracle in favor of a version that is compatible with 0.8.0
* Updated all the contracts to solidity 0.8.13
* Fixed underflow in the vesting contracts + added more tests
* Removed the `pragma experimental ABIEncoderV2` section as it's not required anymore
* Changed some `view` functions to `pure` as now the compiler shows a warning about it
* Removed `public` from constructors  as now the compiler shows a warning about it
* Fixed flaky tests that only happened on M1 machines (basically we were missing `await` before some `approve` transactions

## Any special instructions for the reviewer?

* This is a VERY dangerous PR!! Please go over the changes very carefully. I did my best, but even just one change could break things.
* Please also run the local application from scratch and perform ALL the operations in the system and see that everything is like the expected behavior

kazazor commented:
@AndrzejGisWideLab & @tkeidar the PR is ready for review :)

kazazor commented:
OK Fixed the problem I had locally (we had flaky tests), so had the chance to play around with the system.

It seems to me that there is some problem with our `UniswapV2Router` which we're importing from the `@uniswap/v2-periphery` library.

1. I'm getting an error of `UniswapV2Router: EXPIRED` for swaps and adding liquidity. I did not change anything about this file, but there are two possible problems with it: (A) Something in the uniswap oracle is wrong and it has something to do with it? (B) the file itself works because it's marked as `pragma =0.6.6`, but we're not compiling it but just including its build json files. So maybe something doesn't work here after all? This option is less likely IMO.

2. When I start the application from scratch, the BDX price is now almost $20 and it used to start at $1. I need to check what's going on here as it might say that there is a problem with the Uniswap Oracle if the price is off like that.. But, the stable prices make sense so still need to understand what is going on.

AndrzejGisWideLab commented:
> 1. UniswapV2Router: EXPIRED

@kazazor this is often the case if yo don't initialize the system after deplpoyment

kazazor commented:
Thanks @AndrzejGisWideLab , but I did initialize it.. So maybe something is wrong here based on my changes.

I also went back o master now and BDX does start at a price lower than $1 and not $19.. So something is off here with the oracle/router.

kazazor commented:
OK @AndrzejGisWideLab it seems the problem was not related at all, needed to delete the cache and artifacts + node_modules and only then start over. So please also do it while going over the system on this branch.

Everything is working fine now I tried the flows in the system and it seems like it's working for me.

Also, I found another good source for our new UniswapPairOracle contracts on ConffinFinance (an activefork of Frax on Fanthom) that did the same (their TVL is $400K+):
* [CoffinOracle](https://github.com/CoffinFinance/coffin-contracts/blob/main/contracts/CoffinOracle.sol)
* [Babylonian](https://github.com/CoffinFinance/coffin-contracts/blob/main/contracts/libs/Babylonian.sol)
* [FixedPoint](https://github.com/CoffinFinance/coffin-contracts/blob/main/contracts/libs/FixedPoint.sol)
* [UQ112x112](https://github.com/CoffinFinance/coffin-contracts/blob/main/contracts/libs/UQ112x112.sol)

All of that increased my confidence in the work I did on migrating the uniswap oracle to 0.8.13 🚀 

---------------
@AndrzejGisWideLab & @tkeidar I would like to ask both of you to **try ALL the workflows in the system** (including minting/redeeming when the CR is not 100%). I tried it and seems fine, but since it's such a critical PR I believe that would be worth doing also by the two of you.

If you agree please respond with a 👍

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -177,10 +176,10 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
 
         // Calculates collateral needed to back each 1 BdStable with $1 of collateral at current collat ratio
-        uint256 required_collat_fiat_value_d18 = total_supply.mul(global_collateral_ratio_d12).div(BdPoolLibrary.COLLATERAL_RATIO_MAX);
+        uint256 required_collat_fiat_value_d18 = (total_supply * global_collateral_ratio_d12) / BdPoolLibrary.COLLATERAL_RATIO_MAX;
```
I'll comment on `unnecessary parentheses` once here and then we'll decide how to apply it also to the next ones.

I agree they are not needed, but when I looked at the different lines of code there were some long ones that actually helped me understand what's happening. I think it doesn't really change anything in the code/optimization, but seems more readable to me.

WDYT?

kazazor replied:
I am taking it back actually. I'll remove it in most places, but I'm going to keep it in conditions as IMO it is more readable and I would consider these comments as a NIT :)

kazazor replied:
Well, now that I tried applying the changes, I see that prettier (using the `prettier-plugin-solidity`) is the one adding all these parentheses. So I can't remove any of the parentheses you mentioned in your comments without changing the rule.

So please give me a 👍  to resolve all the `parentheses` comments. If not, please suggest the rule change you would like to apply so we can see how it changes the code.

AndrzejGisWideLab replied:
I connot find such a rule.

kazazor commented on contracts/BdStable/Pools/BdPoolLibrary.sol:
```diff
@@ -1,8 +1,7 @@
+pragma solidity 0.8.13;
 
-import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
```
I agree we can remove it, but for now didn't want to start changing even more in the contracts.. Too much already.
It doesn't harm as it's just a wrapper around the existing regular operators.. SO this could be done as a tech debt somewhere else.

WDYT?

tkeidar replied:
we are already going over the app and testing everything... let's do it now so we won't need to retest everything again on another phase

kazazor replied:
I would recommend doing it in a dedicated PR once this will be merged. Too many changes already.

kazazor replied:
You know what, let me have a look again how big it will be

kazazor replied:
By the way, there is not harm to keep it there to be honest.
If you'll look at its implantation in this version, it does nothing. Just wrapped the native operators.
Sub for example is just:
```solidity
Return a - b
```

tkeidar replied:
Yes, I just didn't understand why to change it specifically in BDStable and not in other files

kazazor replied:
We'll actually I did t change it there SafeMath wasn't used there at all.

What happened here it that BDStable uses the BDStableLibrary which uses SafeMath, so by doing it the code on BDStable was able to "use SafeMath".
That was wrong to begin with and shouldn't be done. While the 0.6 compiler allowed it, the 0.8 did not.
So the code ase did not compile.

I could have fixed it by using SafeMath in the BDstable contract, but thought this is better.

kazazor replied:
OK I've removed all the places we were using SafeMath.
Please take the time to carefully review my changes as some places could be a bit confusing.

kazazor commented on contracts/Staking/Vesting.sol:
```diff
@@ -61,22 +58,27 @@ contract Vesting is OwnableUpgradeable {
-                rewardsToClaim = rewardsToClaim.add(userVestingSchedules[i].totalVestedAmount_d18.sub(userVestingSchedules[i].releasedAmount_d18));
+        uint256 userVestingSchedulesLength = userVestingSchedules.length;
+
+        for (uint256 index = from + 1; index <= to && index <= userVestingSchedulesLength; ++index) {
```
That would not work IMO.
If you'll do that, the next iteration of the loop i would be 1 (due to the ++i).
Which means you'll skip the vesting schedule in index 0 (which you did not check before since you moved the elements in the array).
So basically of the first vesting schedule has fully vested already, and also the second one - you'll skip the second one.

This is why I chose this approach here, not to change the code too much but also cover all cases..

WDYT?

P.S
Regarding the name of the index, I personally hate the use of i and find it less readable :)
But I guess it's a NIT as both of us doesn't care about it too much I believe

kazazor replied:
Actually what I just wrote applies if you would have just not increment the index if it's zero (like continue does), but in your example whenever the first schedule is fully vested we'll stop and not claim the rest up until the `to` index, which is not good as well

kazazor commented on contracts/Governance/Timelock.sol:
```diff
@@ -23,7 +23,7 @@ contract Timelock {
     mapping(bytes32 => bool) public queuedTransactions;
 
-    constructor(address admin_, uint256 delay_) public {
+    constructor(address admin_, uint256 delay_) {
```
https://ethereum.stackexchange.com/a/98457

> Constructors are run only once, when the contract is initially deployed. They can't be called at a later time, so aren't "visible" in the sense that variables or functions are.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -168,7 +167,7 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
 
     function canLegallyRedeem(address who) external view returns (bool) {
-        return block.number.sub(lastMintByUserBlock[who]) >= minimumMintRedeemDelayInBlocks;
+        return (block.number - lastMintByUserBlock[who]) >= minimumMintRedeemDelayInBlocks;
```
Keeping it for conditions: https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841046780

kazazor replied:
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so there is no point in keeping these comments.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -211,20 +210,20 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
         // Step increments are 0.25% (upon genesis, changable)
 
-        if (bdstable_price_cur > price_target_d12.add(price_band_d12)) {
+        if (bdstable_price_cur > (price_target_d12 + price_band_d12)) {
```
Keeping it for conditions: https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841046780

kazazor replied:
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so there is no point in keeping these comments.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -211,20 +210,20 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
+                global_collateral_ratio_d12 = global_collateral_ratio_d12 - bdStable_step_d12;
             }
-        } else if (bdstable_price_cur < price_target_d12.sub(price_band_d12)) {
+        } else if (bdstable_price_cur < (price_target_d12 - price_band_d12)) {
```
Keeping it for conditions: https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841046780

kazazor replied:
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so there is no point in keeping these comments.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -211,20 +210,20 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
+        } else if (bdstable_price_cur < (price_target_d12 - price_band_d12)) {
             //increase collateral ratio
-            if (global_collateral_ratio_d12.add(bdStable_step_d12) >= BdPoolLibrary.COLLATERAL_RATIO_MAX) {
+            if ((global_collateral_ratio_d12 + bdStable_step_d12) >= BdPoolLibrary.COLLATERAL_RATIO_MAX) {
```
Keeping it for conditions: https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841046780

kazazor replied:
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so there is no point in keeping these comments.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -377,16 +376,16 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
 
     function transfer_bdx(address to, uint256 BDX_amount) external onlyOwnerOrPool {
-        require(BDX.balanceOf(address(this)).sub(unclaimedPoolsBDX) >= BDX_amount, "Not enough BDX");
+        require((BDX.balanceOf(address(this)) - unclaimedPoolsBDX) >= BDX_amount, "Not enough BDX");
```
Keeping it for conditions: https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841046780

kazazor commented on contracts/Bdx/BDXShares.sol:
```diff
@@ -43,7 +42,7 @@ contract BDXShares is ERC20Upgradeable, OwnableUpgradeable {
         uint256 amount
     ) public onlyOwner {
-        require(totalSupply().add(amount) <= MAX_TOTAL_SUPPLY, "BDX limit reached");
+        require((totalSupply() + amount) <= MAX_TOTAL_SUPPLY, "BDX limit reached");
```
Keeping it for conditions: https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841046780

kazazor replied:
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so no point in keeping these comments.

kazazor commented on test/integration/vesting.ts:
```diff
@@ -133,6 +133,43 @@ describe("Vesting", () => {
     expect(balanceAfterClaim.sub(balanceBeforeClaim)).to.be.eq(amount_d18.mul(2));
   });
 
+  it("claim should transfer multiple vested schedules at once while the 'to' index is exactly up until the number of vested schedules", async () => {
```
I see from your comments that the test is not clear enough.
I did mean to write `number` there, if you'll look at the sentence again you'll see I'm referring to the `to` as an index, and the number refers to the number of vested schedules.

But this test is not clear enough. Just changed it to reflect what I wanted to test, let me know if that's clearer or not. If it doesn't just resolve the comment :)

tkeidar commented on contracts/BdStable/Pools/BdPoolLibrary.sol:
```diff
@@ -1,8 +1,7 @@
 // SPDX-License-Identifier: MIT
-pragma solidity ^0.6.12;
-pragma experimental ABIEncoderV2;
+pragma solidity 0.8.13;
```
I still see many places we are using SaFeMath... is it by purpose?

kazazor replied:
Duplicated, let's continue the discussion in the original comment.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -89,10 +88,10 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
         BDX = IERC20(_bdx_address);
 
-        bdStable_step_d12 = uint256(BdPoolLibrary.PRICE_PRECISION).mul(25).div(10000); // 12 decimals of precision, equal to 0.25%
+        bdStable_step_d12 = (uint256(BdPoolLibrary.PRICE_PRECISION) * 25) / 10000; // 12 decimals of precision, equal to 0.25%
```
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so no point in keeping these comments.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -89,10 +88,10 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
         global_collateral_ratio_d12 = uint256(BdPoolLibrary.COLLATERAL_RATIO_MAX); // Bdstable system starts off fully collateralized (12 decimals of precision)
         price_target_d12 = uint256(BdPoolLibrary.PRICE_PRECISION); // Collateral ratio will adjust according to the 1 <fiat> price target at genesis
-        price_band_d12 = uint256(BdPoolLibrary.PRICE_PRECISION).mul(50).div(10000); // Collateral ratio will not adjust if between 0.995<fiat> and 1.005<fiat> at genesis
+        price_band_d12 = (uint256(BdPoolLibrary.PRICE_PRECISION) * 50) / 10000; // Collateral ratio will not adjust if between 0.995<fiat> and 1.005<fiat> at genesis
```
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so no point in keeping these comments.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -113,7 +112,7 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
         // bdstable_pools_array.length is limited by addPool function
         for (uint256 i = 0; i < bdstable_pools_array.length; i++) {
-            total_collateral_value_d18 = total_collateral_value_d18.add(BdStablePool(bdstable_pools_array[i]).collatFiatBalance());
+            total_collateral_value_d18 = total_collateral_value_d18 + (BdStablePool(bdstable_pools_array[i]).collatFiatBalance());
```
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so no point in keeping these comments.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -129,7 +128,7 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
         } else revert("INVALID PRICE CHOICE. Needs to be either 0 (BDSTABLE) or 1 (BDX)");
 
-        return weth_fiat_price_d12.mul(BdPoolLibrary.PRICE_PRECISION).div(price_vs_weth);
+        return (weth_fiat_price_d12 * BdPoolLibrary.PRICE_PRECISION) / price_vs_weth;
```
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so no point in keeping these comments.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -159,7 +158,7 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
         uint256 bdStable_total_supply = totalSupply();
         uint256 global_collat_value = globalCollateralValue();
-        uint256 efCR = global_collat_value.mul(BdPoolLibrary.PRICE_PRECISION).div(bdStable_total_supply);
+        uint256 efCR = (global_collat_value * BdPoolLibrary.PRICE_PRECISION) / bdStable_total_supply;
```
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so there is no point in keeping these comments.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -177,10 +176,10 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
 
         if (global_collat_value > required_collat_fiat_value_d18) {
-            return global_collat_value.sub(required_collat_fiat_value_d18);
+            return (global_collat_value - required_collat_fiat_value_d18);
```
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so there is no point in keeping these comments.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -238,7 +237,7 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
         uint256 cr = global_collateral_ratio_d12 > effective_collateral_ratio_d12 ? effective_collateral_ratio_d12 : global_collateral_ratio_d12;
 
-        uint256 expectedBdxValue_d18 = BdPoolLibrary.COLLATERAL_RATIO_MAX.sub(cr).mul(totalSupply()).div(BdPoolLibrary.COLLATERAL_RATIO_PRECISION);
+        uint256 expectedBdxValue_d18 = ((BdPoolLibrary.COLLATERAL_RATIO_MAX - cr) * totalSupply()) / BdPoolLibrary.COLLATERAL_RATIO_PRECISION;
```
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so there is no point in keeping these comments.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -250,10 +249,10 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
         }
 
-        uint256 expectedBdx_d18 = expectedBdxValue_d18.mul(BdPoolLibrary.PRICE_PRECISION).div(bdxPrice_d12);
+        uint256 expectedBdx_d18 = (expectedBdxValue_d18 * BdPoolLibrary.PRICE_PRECISION) / bdxPrice_d12;
```
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so there is no point in keeping these comments.

kazazor commented on contracts/BdStable/BDStable.sol:
```diff
@@ -250,10 +249,10 @@ contract BDStable is ERC20Upgradeable, OwnableUpgradeable {
-        uint256 bdxSupply_d18 = BDX.balanceOf(address(this)).sub(unclaimedPoolsBDX);
-        uint256 effectiveBdxCR_d12 = BdPoolLibrary.PRICE_PRECISION.mul(bdxSupply_d18).div(expectedBdx_d18);
+        uint256 bdxSupply_d18 = BDX.balanceOf(address(this)) - unclaimedPoolsBDX;
+        uint256 effectiveBdxCR_d12 = (BdPoolLibrary.PRICE_PRECISION * bdxSupply_d18) / expectedBdx_d18;
```
Resolving all the `unnecessary parentheses` comments as we're discussing it here:
https://github.com/BlindexDAO/BlinDEX/pull/145#discussion_r841047835

Any decision we'll make will be automatically performed (if any) by the prettier, so there is no point in keeping these comments.

kazazor commented:
@tkeidar / @AndrzejGisWideLab I removed SafeMath, please review everything again - be very careful not to miss anything :)

Also, @AndrzejGisWideLab I addressed all your comments, so we just have some small topics to finalize there and we're done :)

tkeidar commented on contracts/Staking/Vesting.sol:
```diff
@@ -53,30 +48,36 @@ contract Vesting is OwnableUpgradeable {
     }
 
     function claim(uint256 from, uint256 to) external {
+        require(from < to, "Vesting: 'to' must be larger than 'from'");
```
what if `from = to`? to claim only one vesting schedule

kazazor replied:
It's not supported even today on the contract. So I did not want to change it for now in this PR.

To be honest it's not something we'll really use as we'll still send from 0 until let's say 100, so not an urgency changing it

tkeidar replied:
agree

tkeidar commented on test/integration/swaps.ts:
```diff
@@ -44,7 +45,7 @@ describe("Swaps", () => {
     const path = [weth.address, bdeu.address, bdx.address];
 
-    weth.approve(router.address, amountIn);
+    await weth.approve(router.address, amountIn);
```
👏👏

AndrzejGisWideLab reviewed (approved):
LGTM I also clicked through all of the wokrflows and all seems to work just fine

AndrzejGisWideLab commented:
We have removed some imports and usings (SafeMatsh). I'm not 100% how these things live in memory and if they are not going to mess up our contracts in case we upgrade them. We should make sure this changes is harmless before we merge

AndrzejGisWideLab commented:
I tested the system before and after upgrade and found no issues
